### PR TITLE
Added decoupled weight decay for optimizers

### PR DESCRIPTION
**Pull request by kashif**
From the paper: "Fixing Weight Decay Regularization in Adam"

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR!

I think we should at least wait for the paper to be out of peer review (the ICLR 2018 results should be out soon) before making a permanent Keras API change. We've been beaten by this several times in the past.

In the meantime, people who want this feature can use this PR as a reference to add it to an existing optimizer.

**Comment by ozabluda** [comment]
beaten => bitten. AKA "subscribe".

**Comment by AvantiShri** [comment]
Fast.ai is making a big deal about being the only deep learning framework with this fix: http://www.fast.ai/2018/07/02/adam-weight-decay/

“the only deep learning framework that implemented the fix was fastai, using code written by Sylvain. Without broad framework availability, day-to-day practitioners were stuck with the old, “broken” Adam.”

**Comment by kashif** [comment]
@AvantiShri funny... chainer has it for a while already... my pytorch PR is also pending since beginning of the year... anyways, would @fchollet you be ok with a flag that switches between the l2 reg. and decay?
